Add py39 support

skupr-anaconda opened the pull request.

skupr-anaconda committed:
Add py39 support

skupr-anaconda committed:
Skip s390x

skupr-anaconda commented:
An error on s390x and osx-arm64:
```
conda.exceptions.ResolvePackageNotFound:
  - libgpuarray==0.7.6
```
We need to add libgpuarray==0.7.6 on arm64 and skip s390x for now

tobijk reviewed (approved):
LGTM

tobijk commented on recipe/meta.yaml:
```diff
@@ -10,34 +10,40 @@ source:
 build:
-  number: 0
+  number: 1
+  # Anaconda's s390x does not contain depdendencies for libgpuarray 0.7.6, this package is skipped for s390x at this time.
```
Not important, but I suggest phrasing this "libgpuarray 0.7.6 cannot be built on s390x because of other missing dependencies...".

varlackc reviewed (commented):
The `doc_url` seems to be deprecated
http://deeplearning.net/software/libgpuarray/

Maybe we should have the `home` `doc_url` and `dev_url` as `https` instead of `http`

skupr-anaconda committed:
Fix home, dev urls, remove deprecated doc_url

skupr-anaconda commented:
> The `doc_url` seems to be deprecated http://deeplearning.net/software/libgpuarray/
> 
> Maybe we should have the `home` `doc_url` and `dev_url` as `https` instead of `http`

I fixed home and dev urls and removed deprecated doc_url. I didn't find other docs for this package

tobijk commented:
I created a new conda environment into which I installed only libgpuarray. ThenI searched for libgpuarray.dylib and ran

```
(libgpuarray) anaconda@static lib % nm -gU libgpuarray.dylib | grep _GpuArray_clear
0000000000006ae0 T _GpuArray_clear
```

Apparently, the symbol is there, so the question is, why it is not found. Onosx-64, when examining the Python module, we see

```
(base) tkoch@x86_64-apple-darwin13 pygpu % otool -L gpuarray.cpython-37m-darwin.so
gpuarray.cpython-37m-darwin.so:
	@rpath/libgpuarray.3.dylib (compatibility version 3.0.0, current version 3.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

But on osx-arm64 we have

```
(libgpuarray) anaconda@static pygpu % otool -L gpuarray.cpython-39-darwin.so
gpuarray.cpython-39-darwin.so:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

Meaning libgpuarray is not linked. That we need to fix. Let's go back to the build and see, if it was linked:

```
arm64-apple-darwin20.0.0-clang -bundle -undefined dynamic_lookup -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,$PREFIX/lib -L$PREFIX/lib -L$PREFIX/lib -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,$PREFIX/lib -L$PREFIX/lib -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,$PREFIX/lib -L$PREFIX/lib -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/pygpu-0.7.6 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I$PREFIX/include -L$PREFIX/lib -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -mmacosx-version-min=11.1 -arch arm64 build/temp.macosx-11.1-arm64-3.9/pygpu/collectives.o -lgpuarray -o build/lib.macosx-11.1-arm64-3.9/pygpu/collectives.cpython-39-darwin.so
```

Yes. But right below it says

```
ld: warning: ignoring file /Users/anaconda/Workspace/pygpu/pygpu_1642765469663/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/libgpuarray.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
```

Oh no. Let's just quickly see if that is indeed so:

```
(libgpuarray) anaconda@static lib % uname -m
arm64
(libgpuarray) anaconda@static lib % file libgpuarray.dylib
libgpuarray.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

Yes :(

tobijk commented:
It is actually the correct package, that was uploaded, but apparently libgpuarray was built for arm64. Probably the build system needs to be tweaked.

tobijk commented:
Fix is here https://github.com/AnacondaRecipes/libgpuarray-feedstock/pull/2
